### PR TITLE
Use nullifier hash instead of nullifier as public input

### PR DESCRIPTION
**Pull request by drewstone**

### CODE REVIEW

File: pallets/merkle/src/lib.rs [hunk 1]
```diff
@@ -355,19 +357,35 @@ impl<T: Config> Group<T::AccountId, T::BlockNumber, T::GroupId> for Module<T> {
...
 		verifier.constrain(leaf_lc - var_leaf);
+		// committed nullifier into a hash should match hash of nullifier
+		let nullifier_hash_lc = Data::constrain_verifier(
+			&mut verifier,
+			&pc_gens,
+			var_nullifier.into(),
+			var_nullifier.into(),
+			&h
+		);
+		verifier.constrain(nullifier_hash_lc - LinearCombination::from(nullifier_hash.0));
```
Comment (filiplazovic) [inline pallets/merkle/src/lib.rs:388]: We can subtract Scalar from LinearCombination, no need for conversion

File: pallets/merkle/src/merkle/helper.rs [hunk 2]
```diff
@@ -15,16 +15,23 @@ pub struct ZkProof {
...
 	pub path: Vec<(Commitment, Commitment)>,
 	pub r_com: Commitment,
-	pub nullifier: Data,
+	pub nullifier_com: Commitment,
+	pub nullifier_hash: Data,
 	pub bytes: Vec<u8>,
 }
 
+/// Constrain a linear combination to be equal to a scalar
+pub fn constrain_lc_with_scalar<CS: ConstraintSystem>(cs: &mut CS, lc: LinearCombination, scalar: &Scalar) {
```
Comment (filiplazovic) [inline pallets/merkle/src/merkle/helper.rs:24]: If we opt to use this function, then we should use it everywhere, but it doesn't make our life terribly easier

**Review by filiplazovic — APPROVED** [review]

File: pallets/merkle/src/lib.rs (hunk 1 above)
Comment (filiplazovic) [reply pallets/merkle/src/lib.rs:388]: Although bulletproofs is missing some of the arithmetic impls, e.g. for references of LinearCombinations

File: pallets/merkle/src/merkle/helper.rs (hunk 2 above)
Comment (drewstone) [reply pallets/merkle/src/merkle/helper.rs:24]: Agreed, I may end up swapping our gadgets this week so we do use this in gadgets lib.

**Pull request merged**